### PR TITLE
net: sockets: Fix recv() not being interrupted on close()

### LINKED ISSUE #58280: net: sockets: receiving thread not unblocked when closing the blocking socket
**Describe the bug**

I noticed an issue with the condition variable which is used in the socket layer to implement blocking reads. If Thread A calls `recv()` on a blocking socket and Thread B calls `close()` on the same socket, Thread A will keep waiting in `wait_data()` even if the socket is closed.

I did not noticed that issue at first, but this is a problem when reconnecting afterwards. The same socket can be reused but the condition variable is reinitialized in `zsock_socket_internal()` while another thread is still waiting on it.

In order to solve that issue, I signalled the condition variable in `zsock_flush_queue()` to unblock the reader thread, like what is already done in `zsock_received_cb()`:
```
--- a/subsys/net/lib/sockets/sockets.c
+++ b/subsys/net/lib/sockets/sockets.c
@@ -147,6 +147,7 @@ static void zsock_flush_queue(struct net_context *ctx)
 
        /* Some threads might be waiting on recv, cancel the wait */
        k_fifo_cancel_wait(&ctx->recv_q);
+       (void)k_condvar_signal(&ctx->cond.recv);
 }
 
 int zsock_socket_internal(int family, int type, int proto)
```

Is that a proper fix or am I missing something ?

**Expected behavior**

The receiving thread is unblocked when another thread calls `close()` on the blocking socket.

### PR DESCRIPTION
**Pull request by rlubos**
In case recv() call was waiting for data, and the socket was closed from
another thread, the recv() call would not be interrupted, causing the
receiving thread to be blocked indefinitely.

Fix this, by signalling the condvar the recv() call is waiting on
close(). Additionally, close will now set the socket into error mode,
with EINTR as the error condition, allowing the blocked calls to
recognise that the call was interrupted, and return a proper error code
on the event.

Fixes #58280

### CODE REVIEW

File: subsys/net/lib/sockets/sockets.c [hunk 1]
```diff
@@ -154,6 +154,9 @@ static void zsock_flush_queue(struct net_context *ctx)
 
 	/* Some threads might be waiting on recv, cancel the wait */
 	k_fifo_cancel_wait(&ctx->recv_q);
+
+	/* Let reader to wake if it was sleeping */
```
Comment (pdgendt) [inline subsys/net/lib/sockets/sockets.c:158]:
```suggestion
	/* Wake reader if it was sleeping */
```

**Review by jukkar — DISMISSED** [review]
:+1:

**Review by pdgendt — DISMISSED** [review]

**Review by pdgendt — DISMISSED** [review]

**Review by pdgendt — DISMISSED** [review]

**Comment by vaussard** [comment]
Looks good, thank you for the fix :+1: 
I won't be able to test on the hardware until next Tuesday, but this should work.
I wonder if this should not be backported into 2.7 LTS?

**Comment by rlubos** [comment]
@pdgendt Sorry for yet another push, I now realized I returned a non-negative value incorrectly...

**Comment by rlubos** [comment]
> I wonder if this should not be backported into 2.7 LTS?

May be tricky, as the commit won't apply cleanly on 2.7 (socket error flag is not present there yet).

**Review by jukkar — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]

**Pull request merged**